endpoints/openrtb2/auction.go

```diff
@@ -527,34 +529,45 @@ func validateNativeAsset(asset nativeRequests.Asset, impIndex int, assetIndex in
 
 	if asset.Img != nil {
 		if foundType {
-			return fmt.Errorf("request.imp[%d].native.request.assets[%d] must define at most one of {title, img, video, data}", impIndex, assetIndex)
-		}
-		foundType = true
-		if err := validateNativeAssetImg(asset.Img, impIndex, assetIndex); err != nil {
-			return err
+			multipleAssets = true
+		} else {
+			foundType = true
+			if err := validateNativeAssetImg(asset.Img, impIndex, assetIndex); err != nil {
+				return err
+			}
 		}
 	}
 
 	if asset.Video != nil {
 		if foundType {
-			return fmt.Errorf("request.imp[%d].native.request.assets[%d] must define at most one of {title, img, video, data}", impIndex, assetIndex)
-		}
-		foundType = true
-		if err := validateNativeAssetVideo(asset.Video, impIndex, assetIndex); err != nil {
-			return err
+			multipleAssets = true
+		} else {
+			foundType = true
+			if err := validateNativeAssetVideo(asset.Video, impIndex, assetIndex); err != nil {
+				return err
+			}
 		}
 	}
 
 	if asset.Data != nil {
 		if foundType {
-			return fmt.Errorf("request.imp[%d].native.request.assets[%d] must define at most one of {title, img, video, data}", impIndex, assetIndex)
-		}
-		foundType = true
-		if err := validateNativeAssetData(asset.Data, impIndex, assetIndex); err != nil {
-			return err
+			multipleAssets = true
+		} else {
+			foundType = true
+			if err := validateNativeAssetData(asset.Data, impIndex, assetIndex); err != nil {
+				return err
+			}
 		}
 	}
 
+	if multipleAssets {
+		return fmt.Errorf("request.imp[%d].native.request.assets[%d] must define at most one of {title, img, video, data}", impIndex, assetIndex)
+	}
+
+	if !foundType {
+		return fmt.Errorf("request.imp[%d].native.request.assets[%d] must define at least one of {title, img, video, data}", impIndex, assetIndex)
+	}
```

Comment:
The introduction of multipleAssets in this function extends execution of the function beyond when it might have returned in the function's original form with no change to the result.  Do we really need to do things like this just to satisfy a coding tool?

The final check for foundType changes the behavior of this function from before.  Does that mean the original code has a bug?

Reply:
Do you mean that with the introduction of `multipleAssets` we're unnecessarily checking for each asset and then return when we could have just returned when we saw the first multipleAsset? If so, yeah, you make a fair point and well, this is not just to satisfy the tool but to make the code simpler. I guess a better alternative would be to define the error string once and then reuse that rather than checking `multipleAssets` or using the same error string again and again.

I will update the PR. Thanks for pointing it out

Reply:
Yes, that's what I meant.  Probably not a big deal since it's an error case and, hopefully will happen very rarely.  My bigger concern was related to my second comment, as that introduces behavior that wasn't present before.  Why is that being done?

Reply:
Yeah, the other check is basically making sure that there's at least one native asset present of {title, image, video, data}. With none of them present, the asset object should be considered as invalid. IAB Spec reference: https://www.iab.com/wp-content/uploads/2018/03/OpenRTB-Native-Ads-Specification-Final-1.2.pdf#page=13.

Reply:
I was a little concerned that, IMHO, the wording in the above spec was ambiguous, especially given the hitherto code support of no value being present and the wording of the PBS error messages.   I was able to confirm that other company products also expect exactly one of the four values to be present.  This leads to one final suggestion:  The new multipleAssetErr error message mentions "at most one of" but the new foundType message mentions "at least one of".  I propose that both should say "exactly one of" (which is what I would have liked to see in the spec, as well).

Reply:
Updated @josephveach